feat(backend): Refactor authz to perform SubjectAccessReview. Fixes #3513

elikatsis opened the pull request:
**Description of your changes:**

With this PR we implement the refactoring described in #3513.

In the past, KFP used KFAM for its authorization checks, checking only if the user can `view` the corresponding namespace.
We now refactor the authorization mechanism to perform proper and fine-grained authorization via K8s RBAC and SubjectAccessReview.

Closes #3513 

/assign @Bobgy 

cc @yanniszark 

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 
<!--
   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.
-->
- [x] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).
<!--
    **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update before release.
-->

Bobgy commented on backend/src/apiserver/client/subject_access_review_fake.go:
```diff
@@ -0,0 +1,69 @@
+// Copyright © 2020 Arrikto Inc.  All Rights Reserved.
```
It's okay to keep Arrikto copyright, but please add apache license header to every file.

elikatsis replied:
Oops, I forgot this! Thanks for catching it

Bobgy commented on backend/src/apiserver/client/subject_access_review_fake.go:
```diff
@@ -0,0 +1,69 @@
+func (FakeSubjectAccessReviewClient) Create(*authzv1.SubjectAccessReview) (*authzv1.SubjectAccessReview, error) {
+	return &authzv1.SubjectAccessReview{Status: authzv1.SubjectAccessReviewStatus{
+		Allowed:         true,
+		Denied:          false,
```
When will denied be true, what does it mean?

elikatsis replied:
From the [documentation](https://godoc.org/k8s.io/api/authorization/v1#SubjectAccessReviewStatus):
```go
type SubjectAccessReviewStatus struct {
    // Allowed is required. True if the action would be allowed, false otherwise.
    Allowed bool `json:"allowed" protobuf:"varint,1,opt,name=allowed"`
    // Denied is optional. True if the action would be denied, otherwise
    // false. If both allowed is false and denied is false, then the
    // authorizer has no opinion on whether to authorize the action. Denied
    // may not be true if Allowed is true.
    // +optional
    Denied bool `json:"denied,omitempty" protobuf:"varint,4,opt,name=denied"`
    ...
}
```
So I believe we only care about the `Allowed` field to decide.
Maybe if we had multiple authorization mechanisms **and** we found out that `allowed == denied == false`, then we should proceed with the next mechanism.

Bobgy replied:
SGTM, thanks for the info!

Bobgy commented on backend/src/apiserver/client/subject_access_review.go:
```diff
@@ -0,0 +1,48 @@
+func createSubjectAccessReviewClient() (v1.SubjectAccessReviewInterface, error) {
+	restConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to initialize kubernetes client.")
```
When failing with in cluster config, can we try kubeconfig, so that we can run API server locally?

elikatsis replied:
If I understand correctly, this is the same with kubernetes core client initialization ([source](https://github.com/kubeflow/pipelines/blob/a8bd4682656e75b5c5cd48d75ffdf23e97bf99aa/backend/src/apiserver/client/kubernetes_core.go#L25-L36))

This means we should patch both places (and every other place) if we want to run it locally.
Would it be better to file an issue and have a distinct PR for it?

Bobgy replied:
Thanks, I wasn't aware of the existing convention, makes sense as separate issue and PR.

elikatsis replied:
I created an issue to track the effort: https://github.com/kubeflow/pipelines/issues/4738
So I'm marking this as resolved as well

Bobgy commented on backend/src/apiserver/client/subject_access_review.go:
```diff
@@ -0,0 +1,48 @@
+	var err error
+	var operation = func() error {
+		client, err = createSubjectAccessReviewClient()
+		if err != nil {
+			return err
+		}
+		return nil
```
nit: this is equivalent to 

return err

elikatsis replied:
I blindly copied this part from the kubernetes core client :smile:. I'll do this fix in both places

Bobgy reviewed (commented):
Thank you so much @elikatsis for the contribution! I have been looking forward for this for a while!

Some early feedback first. I haven't gone through everything.

Bobgy commented:
@elikatsis I've got one general question, what would be the replacement you think for KFAM for namespace discovery?
Previously, we relied on centraldashboard talking to KFAM to get that, but I believe we can no longer do so if we completely remove KFAM.

I'm thinking about listing all namespaces with a certain pipelines.kubeflow.org/xxx label and build a new selector using this info collected in backend. I wonder what do you think about it.

/cc @yanniszark

elikatsis commented:
/retest

Bobgy commented:
I was kind of busy, will review tomorrow

Bobgy commented on backend/src/apiserver/client/subject_access_review_fake.go:
```diff
@@ -0,0 +1,81 @@
+	}}, nil
+}
+
+func (FakeSubjectAccessReviewClient) CreateContext(context.Context, *authzv1.SubjectAccessReview) (*authzv1.SubjectAccessReview, error) {
```
Have you considered making a thin interface for this client, so that we do not need to expose this method we don't use.
example: https://github.com/kubeflow/pipelines/blob/1165ed243b381ae37a72e245d685ba318d58b345/backend/src/apiserver/client/argo.go#L28-L34

chensun replied:
Actually where does this `CreateContext` come from? I don't see it being asked by the interface:
https://godoc.org/k8s.io/client-go/kubernetes/typed/authorization/v1#SubjectAccessReviewInterface

elikatsis replied:
We are using client-go v0.17.9: https://pkg.go.dev/k8s.io/client-go@v0.17.9/kubernetes/typed/authorization/v1#SelfSubjectAccessReviewInterface

I wrote a custom `SubjectAccessReviewInterface` which we can remove it in the future, when we upgrade `client-go`.

Bobgy reviewed (commented):
Thank you so much @elikatsis!
I just read through all the code, left some nit pickings. The high-level structure looks awesome to me. Let's follow up with some details and I think we'll be good to go.

Another general nit picking I've seen a few times is that for each resource we defined a `canAccessXXX(id)` method that takes the ID, so that we needs to first fetch this resource to know its namespace. In cases that we already fetch this resource and know its namespace, we wrote one-off boilerplate code to fetch the resource and check authorization.

I think we can simplify a little bit by defining two utility methods for each resource:
* `canAccessXxxById(id, verb)` --- fetches the resource inside and figures out namespace/name
* `canAccessXxx(namespace, name, verb)` --- we are already fetching the resource, so we can pass namespace and name directly

`canAccessXxxById` will be straightforward to implement using `canAccessXxx` and we can avoid quite some duplication there.

Bobgy commented:
Other than that, can you resolve the conflicts?

maganaluis commented:
We are working on https://github.com/kubeflow/pipelines/issues/4197 and I just realized our changes will also have conflicts with this PR. Just for my own planning, will this be merged for the release 1.3? What else will we need to integrate with SubjectAccessReview. I know the CentralDashboard also uses KFam. 

@elikatsis @Bobgy

Bobgy commented:
Yes, I think as long as @elikatsis gets through the comments, we can merge.

It should be ready with kubeflow 1.3

This change is compatible with central dashboard, profile controller and kfam, because profile controller should already give service accounts rolebindings that can aggregate ones in this PR.

Bobgy commented:
It'll be ideal if we can get rid of kfam from central dashboard too, so that we can completely remove it from the manifests, but there are still some unanswered design questions.

elikatsis commented on manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml:
```diff
@@ -0,0 +1,108 @@
+  clusterRoleSelectors:
+  - matchLabels:
+      rbac.authorization.kubeflow.org/aggregate-to-kubeflow-pipeline-edit: "true"
+rules: []
```
@Bobgy,
Great question! The answer is this essentially: https://github.com/kubernetes/kubernetes/issues/65171

TL;DR: We can't have both aggregation **and** rules in a [Cluster]Role. The rules will be ignored. This is why we need to define our rules in some [Cluster]Roles and then have _one role to find them, one role to bring them, and in itself bind them_ :smile:

Bobgy replied:
Thanks for the explanation! Good to know. : )

elikatsis commented on manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/pipeline.yaml:
```diff
@@ -273,6 +273,12 @@ rules:
       - update
       - patch
       - delete
+  - apiGroups:
```
Done

elikatsis commented on backend/src/apiserver/server/job_server.go:
```diff
@@ -88,11 +89,37 @@ func (s *JobServer) CreateJob(ctx context.Context, request *api.CreateJobRequest
-	err = CanAccessExperimentInResourceReferences(s.resourceManager, ctx, request.Job.ResourceReferences)
 	if err != nil {
 		return nil, util.Wrap(err, "Failed to authorize the request.")
 	}
```
Yes! Thanks for catching it

elikatsis commented:
@Bobgy 
I rebased and force-pushed. The minor nits (gcp marketplace manifests and forgotten double error check) are included in the "old" commits.
For the comments about
* centralized check for multi-user shared mode
* use `canAccessXXX()` instead of `isAuthorized()`
* enhancement on error assertion wrt unauthorized calls
I have added two fixup commits. I will squash them if they look good to you.

As far as having general `canAccessXXX()` and `canAccessXXXById()` functions, I think we don't need them. This is covered by each server's methods (see first `fixup!` commit). Implementing them in a different place and have the server use them looks like it adds complexity instead of simplifying things.

elikatsis commented on backend/src/apiserver/server/experiment_server.go:
```diff
@@ -137,7 +147,16 @@ func (s *ExperimentServer) ListExperiment(ctx context.Context, request *api.List
+			Subresource: "",
+			Name:        "",
+		}
+		err = isAuthorized(s.resourceManager, ctx, resourceAttributes)
```
Done

elikatsis commented on backend/src/apiserver/server/auth_server_test.go:
```diff
@@ -94,7 +94,7 @@ func TestAuthorizeRequest_Unauthorized(t *testing.T) {
 	_, err := authServer.Authorize(ctx, request)
 	assert.Error(t, err)
-	assert.EqualError(t, err, "Failed to authorize the request: Failed to authorize namespace: BadRequestError: Unauthorized access for user@google.com to namespace ns1: Unauthorized access for user@google.com to namespace ns1")
+	assert.Contains(t, err.Error(), "Unauthorized access")
```
Done (2nd fixup commit). There's quite some refactoring there so please tell me what do you think :)

elikatsis commented on backend/src/apiserver/client/subject_access_review.go:
```diff
@@ -0,0 +1,57 @@
+	err = backoff.Retry(operation, b)
+
+	if err != nil {
+		glog.Fatalf("Failed to create pod client. Error: %v", err)
```
Fixed

elikatsis commented on manifests/kustomize/base/pipeline/cluster-scoped/view-edit-roles.yaml:
```diff
@@ -0,0 +1,108 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
Sure, we will get the manifests repo updated accordingly as soon as we get this merged.
I created the directory you mentioned and moved the role declarations there.

Bobgy replied:
Thank you! That would be great

elikatsis commented on backend/src/apiserver/server/util.go:
```diff
@@ -289,110 +290,49 @@ func getUserIdentityFromHeader(userIdentityHeader, prefix string) (string, error
 	if common.IsMultiUserMode() == false {
 		// Skip authz if not multi-user mode.
 		return nil
 	}
```
I think I covered this with the `fixup!` commits. PTAL

Bobgy commented on backend/src/apiserver/server/util.go:
```diff
@@ -289,110 +290,55 @@ func getUserIdentityFromHeader(userIdentityHeader, prefix string) (string, error
 	}
 
-	glog.Infof("Authorized user %s in namespace %s", userIdentity, namespace)
+	glog.Infof("Authorized user '%s': %+v", userIdentity, resourceAttributes)
```
These authorization logs will be very common, every request comes with it.
Shall we change log level to `trace` or `debug`?

For most of the cases, users should depend on error message for debugging, rather than api server logs.

elikatsis replied:
I can only see `Info`, `Warning`, `Error`, and `Fatal` levels. Can you help me how can I do `debug` or sth different?
Here's what I found in the repo, not sure if it helps: https://github.com/golang/glog/blob/23def4e6c14b4da8ac2ed8007337bc5eb5007998/glog.go#L91-L115

Bobgy replied:
OK, sorry I thought there were more. Then let's go with info for now.

Bobgy commented on manifests/kustomize/base/pipeline/multi-user/view-edit-roles.yaml:
```diff
@@ -0,0 +1,108 @@
+apiVersion: rbac.authorization.k8s.io/v1
```
@elikatsis Can you add an inline comment with short explanation? I think upcomers will be confused too, so it'll be great as a comment.

Bobgy reviewed (commented):
I've only got two comments left. Everything else looks great to me!
Thank you so much @elikatsis!

/lgtm

elikatsis commented:
@Bobgy I added the comment and squashed the fixups. If you can help me selecting the logging level, we'll be golden!

Bobgy commented:
Do you need @yanniszark 's review?
Everything good on my side

elikatsis commented:
@Bobgy we have already reviewed this internally, so we should be good to go. Thank you very much for your comments! :smiley:

Bobgy commented:
/lgtm
/approve
Cool, I am feeling excited about the huge potential this brings. Thank you again for this great contribution!
